include/extractor/guidance/intersection.hpp

```diff
@@ -215,10 +222,51 @@ struct IntersectionView final : std::vector<IntersectionViewData>,      //
     using Base = std::vector<IntersectionViewData>;
 };
 
+// `Intersection` is a relative view of an intersection by an incoming edge.
+// `Intersection` are streets at an intersection ordered from from sharp right counter-clockwise to
+// sharp left where `intersection[0]` is _always_ a u-turn
+
+// An intersection is an ordered list of connected roads ordered from from sharp right
+// counter-clockwise to sharp left where `intersection[0]` is always a u-turn
+//
+//                                           |
+//                                           |
+//                                     (intersec[3])
+//                                           |
+//                                           |
+//                                           |
+//  nid ---(via_eid/intersec[0])--- nbg.GetTarget(via)  ---(intersec[2])---
+//                                           |
+//                                           |
+//                                           |
+//                                     (intersec[1])
+//                                           |
+//                                           |
+//
+// intersec := intersection
+// nbh := node_based_graph
+//
 struct Intersection final : std::vector<ConnectedRoad>,         //
                             EnableIntersectionOps<Intersection> //
 {
     using Base = std::vector<ConnectedRoad>;
+
+    /*
+     * Check validity of the intersection object. We assume a few basic properties every set of
+     * connected roads should follow throughout guidance pre-processing. This utility function
+     * allows checking intersections for validity
+     */
+    bool valid() const;
+
+    // given all possible turns, which is the highest connected number of lanes per turn. This value
+    // is used, for example, during generation of intersections.
+    std::uint8_t getHighestConnectedLaneCount(const util::NodeBasedDynamicGraph &) const;
+
+    // check if all roads between first and last allow entry
+    bool hasValidEntries(Intersection::iterator first, Intersection::iterator last) const;
+
+    Base::iterator findClosestTurn(double angle);
+    Base::const_iterator findClosestTurn(double angle) const;
```

Comment:
All of the functions here should no longer be attached to `Intersection` or `IntersectionView` - but instead check `EnableIntersectionOps<T>`. If you have the time for it, read up about the CRTP pattern.

We need this since even though this creates a inheritance hierarchy

```c++
struct base {};
struct derived : base {};
```

the following types stand in no relationship to each other

```c++
using BaseVec = vector<base>;
using DerivedVec = vector<derived>;
```

using the CRTP pattern we inject the functionality into both collection types.